### DISCUSSION TITLE
Question: Recommended settings to speed up scroll performance?

### DISCUSSION THREAD

**Discussion by inthreedee** [photoprism/photoprism#1036] [Q&A] [answered]
Are there any recommended settings to help speed up scrolling through a large library?

I'm currently seeing a 2-3 second wait time to load the next group of photos after reaching the bottom of the page on an 8GB Raspberry Pi 4.  It doesn't appear to be taxing the CPU much at all, and it's running off a decent SSD so I don't think that would be the bottleneck.  The thumbnails it's loading are only around 25kb each.  Could there be another bottleneck I'm not thinking of?  Or is there a setting to be more aggressive about loading more photos before reaching the bottom of the page?

Just looking for any advice on what I could tweak to make scrolling a bit more smooth.  Thanks!

**Comment by lastzero** [comment, marked as answer]
There may be various reasons, see #1014: The database, the browser, the number of http connections can be blocking as well. Of course you can make preloading even more aggressive, but then you might experience more of a lag while scrolling as JS is single threaded - so scrolling & loading at the same time is not ideal. You can also preload more files at once, but then your browser may be blocking as too many HTTP connections are being used. Cloud apps are using multiple domains for their CDN, which we can't. Using a CPU with more cores might help as well (even though the load may be low as you mentioned).

Note that the Mosaic view wasn't really built for scrolling through very fast like the timeline view that is on our roadmap. If you slow down a little bit, it should work just fine.

**Reply by inthreedee** [reply]
It sounds like the upcoming timeline view is going to provide a solution. Thanks, I appreciate the info!

**Reply by miniminime** [reply]
> There may be various reasons, see #1014: The database, the browser, the number of http connections can be blocking as well. Of course you can make preloading even more aggressive, but then you might experience more of a lag while scrolling as JS is single threaded - so scrolling & loading at the same time is not ideal. You can also preload more files at once, but then your browser may be blocking as too many HTTP connections are being used. 

I have here PhotoPrism running with 16GB RAM, 8/16 CPU cores (AMD Ryzen 7 1700) on Ubuntu 20.04 and having the same "sluggishness". I tried the latest Firefox and Chrome browser on Windows and the result is the same. The next pictures are only loaded when you really reach the bottom of the page and then it takes 1-2 seconds until the new pictures are being displayed.

Maybe it's possible to load the next batch of pictures just a bit earlier? An additional parameter to influence this would be quite cool as I understand not everyone has the needed horsepower, but for those who do it would be nice to make use of it.

**Reply by lastzero** [reply]
When this happens, you are either scrolling down very far - or you have many, many photos and the current query has a sort order or filter that is hard to optimize for the database (note that Google Photos doesn't even allow sorting for this reason).

In all other cases, there could be other issues like I/O problems or database file locking when using SQLite.

We understand it's better for some users and use cases to just keep scrolling instead of refining search filters or use other views, like the calendar. That's why providing a Timeline View ison top of our Roadmap: https://github.com/photoprism/photoprism/issues/152

You just need to give us a bit more time and ideally become a sponsor. Thank you! ✨

**Reply by lastzero** [reply]
Check out our map view in *Places*: It loads tens of thousands of files within a second!

![Screenshot 2021-12-01 at 16 38 28](https://user-images.githubusercontent.com/301686/144264941-4641543b-68da-40f1-bf24-e8189a789c52.png)

**Reply by miniminime** [reply]
Thanks for the reply!

No filter applied, not that many photos at the very moment (only 600), just from the start page. I am using a PCIe SSD, that shouldn't be an I/O bottleneck either. DB is MariaDB.

I don't want to make a big fuss out of this, just wanted to provide a bit more information. Happy to wait for #152, I just thought maybe there is an easy fix or config trick to remedy this a bit.  :-)

And yes, I will consider to become a sponsor. I thought of that already in fact. I understand that you prefer continuous stream of donations, but could it be an idea to accept donations for specific features as a one-off as well?

**Reply by inthreedee** [reply]
> When this happens, you are either scrolling down very far - or you have many, many photos and the current query has a sort order or filter that is hard to optimize for the database

Just pointing out that this happens on the demo site too, https://demo.photoprism.org/browse
Load it up, scroll down to the bottom, and notice that it takes 1-2 seconds to load the next batch.  Scroll again, same result.  Then you're at the end of the photos available in the demo, so it's not that there are too many photos there.

It seems like this is just normal behavior for that view.  People are used to scrolling quickly through photos, and the default view simply doesn't support the kind of behavior those users expect.  It's okay to admit that, since timeline view is on its way for an improved experience!

**Reply by lastzero** [reply]
Yes, it's true, the normal views are not designed for fast scrolling. If you scroll down in a straight line, then the current trigger point is not aggressive enough. I wouldn't consider this a "performance" issue, and it's not related to the browser, server, or client device in this case.

**Reply by lastzero** [reply]
@miniminime GitHub Sponsors now also offers one-time donations: https://github.com/sponsors/photoprism?frequency=one-time

**Reply by lastzero** [reply]
Note that the demo is hosted in Germany. So testing from other parts of the world may result in a different experience than e.g. from our office in Berlin:
- https://github.com/photoprism/photoprism/issues/1351

**Reply by miniminime** [reply]
> Yes, it's true, the normal views are not designed for fast scrolling. If you scroll down in a straight line, then the current trigger point is not aggressive enough. I wouldn't consider this a "performance" issue, and it's not related to the browser, server, or client device in this case.

Final thought on this and then I let it go, promised  :-D

What if instead of having a fixed trigger-scrolling-point you would rather continuously having x-number of "unseen" pictures loaded at the end of the page so that there is a continuous flow of pictures? This could remove the perceived sluggishness, I think, and wouldn't burden the systems to much as it stops loading further if you don't look at the "unseen" pictures. I understand it's kind of the same as if you would just place the trigger a bit earlier (could I change this myself, somewhere in the code?)....

**Reply by lastzero** [reply]
Yes, this is a possible optimization. We're not running out of ideas here, just resources, and we've decided to focus on multi-user support at the moment.

**Reply by lastzero** [reply]
In fact, the frontend already preloads results for the photo viewer, which is a waste of resources and needed because it can't dynamically load more. Our strategy is to improve / replace the viewer first, as this automatically optimizes search performance and removes complexity:

- https://github.com/photoprism/photoprism/issues/1307

We can then proceed with further optimizations without maintaining the current workaround and coding around it.

**Reply by inthreedee** [reply]
@lastzero Just wanted to pop back in here and say that your performance improvements this year make a world of difference. Hosted on my RPi4, scrolling is now nearly as fast and smooth as google photos.

Well done, and much appreciated! :heart:

**Comment by miniminime** [comment]
I just observed something: When you go to a calendar album or also from the start page it will tell you that it found more than 50 photos for a brief moment. If you then on the top right click on "reload" you get immediately more pictures and the continued loading seems also faster.

**Reply by lastzero** [reply]
Yes, because it shows the previously loaded number of photos. This is also necessary when you go back from another page to restore the scroll position. Again, it's a tradeoff between many options, and given enough time and feedback we can tweak the usability further.

**Reply by lastzero** [reply]
If you have a lot of photos, it may be helpful to increase the MariaDB cache if you find that search results load faster when you reload them.

**Reply by miniminime** [reply]
Thank you, I will try that.

**Comment by joachimtingvold** [comment]
For all API queries with somewhat static content (most pages with no search/query, albums, etc), the API response could be cached via varnish or similar (and just invalidate the cache upon changes). That should significantly increase the performance, regardless of the database size. Even hitting on the "Search" button (i.e. show all photos) is what I'd call slow in the current state (2-3+ seconds).

I'm coming from a customized version of [PhotoFloat](https://git.zx2c4.com/PhotoFloat/about/). It uses statically generated JSON files (so not comparable to PP in current state), but if caching was implemented in PP, the performance for getting the JSON data would be more or less the same.

In PhotoFloat I could have 2k+ images in the same album, with all spacing for all 2k+ thumbnails being being instantly preloaded on the same page (so the scrollbar would be correct from the get-go), and it would start loading the thumbnails that was in the current view (so scrolling would change what tumbnails where being preloaded). That way I could quickly find the pictures I was looking for (if I knew approximately where in the album those images where).

There was never any issue with blocked connections or performance issues. Everyhing was more or less instant. I'm not complaining about PP, but I must say it was a big difference in that regard coming from PhotoFloat.

I ultimately switched due to all the other features of PP (location/GPS, ML, facial recognition, etc).

**Reply by lastzero** [reply]
Seems wrong to statically cache a dynamic search API, although that obviously makes simple reloads using the same parameters faster (and the result outdated if someone else updates photos at the same time). We could even do this directly in Go, but there are better ways to improve UX without only optimizing very simple use cases. As always, we can't work on everything at the same time. Decided to do facial recognition first, so that's what we have in store.

Even if it takes 2 or 3 seconds until the last request has finished (visible images at the top load first, so the perceived load time is much faster), that's not a lot of time to load 400 results and 230 images from a 100k file library running on an old 2-core Intel Core i3.

The current viewer implementation is a huge pain by the way, without it way less backend requests would be needed. Since it's loading in the background, you may see its "slow" requests in dev tools, but this is not the "real" result load time after you click the search button.

Besides, I don't really get how people can rush from one photo to the next or start a new search every few seconds. My brain is the slowest part here, maybe it's because I'm getting old! 🤣

**Reply by joachimtingvold** [reply]
> Seems wrong to statically cache a dynamic search API, although that obviously makes simple reloads using the same parameters faster (and the result outdated if someone else updates photos at the same time).

As I said, it would only make sense for things that are fairly static; blank search (i.e. "all photos"), albums, etc. Any changes would obviosly have to invalidate the relevant caches. I have not really analyzed where the bottleneck(s) are in PP, but I have my PhotoFloat and PP instances running on the same hardware (16C/192GB RAM, DB+OS/Docker on SSD, originals on HDD) and they are in two different leagues right now (especially the scrolling/preloading part).

> As always, we can't work on everything at the same time. Decided to do facial recognition first, so that's what we have in store.

I'm not expecting you to fix it (-: I have zero expectations of you to fix anything that I bring to the table, and I'm sorry that you repeatedly have to "excuse" yourself to users of PP that you cannot adress all the issues alone at the same time. You shouldn't have to do that, IMHO.

I'm not really complaining (even though it might seem so); I'm just coming with a different perspective to confirm that PP is indeed somewhat "sluggish" compared to other products in this regard.

> Even if it takes 2 or 3 seconds until the last request has finished (visible images at the top load first, so the perceived load time is much faster)

It's showing the black progress bar (and no results/white page) for 2-3 seconds on my setup, before it starts to load anything at all.

> that's not a lot of time to load 400 results and 230 images from a 100k file library running on an old 2-core Intel Core i3.

For relatively "static" content, it is (unless you're actually doing a search). PhotoFloat does it near instantly (JS+JSON, principally the same as PP), on the same hardware, so they are on equal footing.

> Besides, I don't really get how people can rush from one photo to the next or start a new search every few seconds. My brain is the slowest part here, maybe it's because I'm getting old! 🤣

Haha, I don't need to do searches all the time, but the "slowness" is not just related to search. OP was pointing to "slow scroll", which I agree is the most "annoying" part of the slowness (especially compared to how PhotoFloat does it). 50 photos is not really much, and having to wait 1-3 seconds every time you've scrolled past 50 photos... For reference, that's not even a full page on high resolution monitors.

**Reply by lastzero** [reply]
No worries. Feeling confident we'll address most of this with the timeline view and a new full screen viewer.

Obviously PhotoPrism loads lots of CSS & JS initially, so a full reload will always be slow compared to small, static pages. Since it's a large single page app which also initializes a websocket connection for push notifications, that shouldn't surprise anyone. Samewise, vanilla JS would be faster than VueJS + Vuetify.

While there certainly are faster and smaller galleries, the observed performance seems completely reasonable for the type of application and technology.

It's really hard to get a consistent picture on the performance topic. To my big surprise, there are many Raspberry Pi users who say it's comparable to Google Photos while others like you with a 192 GB power server say it's sluggish.

It's reasonable to assume many of the problems reported are really local configuration issues. For example, slow loading and broken thumbnails recently turned out to be a rate limit in a Traefik proxy. Other users have put their database files and thumbs on network storage or have super slow NAS devices and no SSD.

**Reply by lastzero** [reply]
Everyone is invited to share screen videos as this gives a visual impression and enables us to provide more helpful feedback. In some cases, there may be an easy fix, like increasing the database cache or changing proxy / CDN settings.

**Reply by lastzero** [reply]
Oh, and always verify you haven't disabled the browser cache or enabled throttling in dev tools. Happened more than once! 😜

**Reply by inthreedee** [reply]
> Besides, I don't really get how people can rush from one photo to the next or start a new search every few seconds. My brain is the slowest part here, maybe it's because I'm getting old! 

Just wanted to provide a use-case where this is common to give you a better understanding of what we hope Timeline view will eventually let us do:

My typical usage is I'm scanning my photos looking for a particular image or set of images I have in mind.  Usually it's not something I can search for because I don't know the exact date or location, but I'll recognize it when I see other photos I took around the same time.  So I'm scrolling quickly trying to find the right set of photos.

The perceived slowness comes from having to wait for each set of photos to load in one set at a time, not being able to quickly scroll ahead without waiting for photos to load.  For example, let's say I scroll to a certain spot and realize that, nope, the photos I'm looking for were taken at least a week after these photos.  If I took a thousand photos that month, I have a lot of waiting to do to scroll ahead a week or more.

The way Google Photos solves this (and what I hope can happen with PhotoPrism's timeline view), is scrolling is smooth and continues even if the photos aren't fully loaded.  Empty gray boxes show up in place of the unloaded images.  That way, if you're trying to get to the middle of say, June, you can quickly scroll there without waiting for all the early June photos to load in.

Here's a youtube video I found from one of the google engineers demonstrating this scroll behavior: https://youtu.be/d57mzcSrSQw?t=26
And their medium post describing how they helped build the UI: https://medium.com/google-design/google-photos-45b714dfbed1
(Note: I fully realize a lot of what is in that post is probably out of scope for PhotoPrism.  I just thought it was an interesting read.)

I hope that makes sense, and I'm super excited for when you're able to start working on timeline view!

**Reply by lastzero** [reply]
Yep, understood. We developed this for a single-page app with 13 Mio users in 2008/2009. Years before Google.

**Reply by joachimtingvold** [reply]
> It's really hard to get a consistent picture on the performance topic. To my big surprise, there are many Raspberry Pi users who say it's comparable to Google Photos while others like you with a 192 GB power server say it's sluggish.

I can imagine :D It also comes down to how perceptive the users are to such things (some don't care at all), and also how patient (or impatient :D) they are. I'm pretty impatient to things that is "waiting just for the sake of waiting" when I know I don't have to (i.e. I always travel off-season to avoid queues and whatnot).

> Everyone is invited to share screen videos as this gives a visual impression and enables us to provide more helpful feedback.

I'm planning to (-: Just have to get done migrating into PP. Then I'll set up development environments so that I can debug, and potentially help fix stuff.

@inthreedee wrote:
> The way Google Photos solves this (and what I hope can happen with PhotoPrism's timeline view), is scrolling is smooth and continues even if the photos aren't fully loaded.

Yep, except the "date slider", that's basically identical to what PhotoFloat does as well.

**Comment by joachimtingvold** [comment]
I haven't done much benchmarking of Go+MariaDB before (so I don't know what would be reasonable numbers), but trying to fetch info about all photos I currently have imported (only about 67k) via the API (`/api/v1/photos?merged=true&count=1000`) takes almost 5 minutes and 67 requests (since there is a limit of 1k per request);

```
jocke@keklolwtf: ~/photoprism $ time ./get_pp_photos.py
Fetching PP photos from API
Writing PP photos to cache file pp_photos
Found 67366 PP photos

real	4m58.059s
user	0m13.827s
sys	0m1.409s
```

Each request takes about 2-3s in the backend log;

```
photos: found 1,000 results for dist:20 count:1000 merged:true [2.497006551s]
```

And the actual requests takes 4 seconds (with about 1.7MiB of data to download);

```
jocke@keklolwtf: ~/photoprism $ time curl -o testlol.txt -s -H "X-Session-ID: <snip>" 'https://photoprism.foo.bar/api/v1/photos?count=1000&offset=0&merged=true'

real	0m3.929s
user	0m0.026s
sys	0m0.032s
jocke@keklolwtf: ~/photoprism $ ls -alh testlol.txt
-rw-r--r--  1 jocke  staff   1.7M Jan  4 22:51 testlol.txt
```

I'd say that is on the "slow side of things". Trying to do anything reasonable with the API was not really feasible with those numbers (to assign photos to albums based on my old photo gallery), so I had to create a local cache with the data, and use that.

**Reply by joachimtingvold** [reply]
I guess I can do the requests in parallel to save some time, but yeah. The "max 1000 per request via API"-limit also poses som other minor caveats. For example that any changes (i.e. adding or removing photos) during those 67 requests will offset the photos, causing some duplicated data. That actually caused some issues for me (i.e. I had to wait until the current import was done). That's maybe not a problem for a "one user PP" scenario, but even then you could get into issues where your phone is doing auto-sync or similar.

**Reply by inthreedee** [reply]
You imply the work is already done, but I figured I'd plop this here in case it's useful to you.  This uses photo hash or name matching rather than offsets to match up photos with a Google Takeout and add them to albums:
https://github.com/inthreedee/photoprism-transfer-album

**Reply by lastzero** [reply]
Why not use SQL directly? The current API has a very specific use case, which definitely isn't mass export. Besides 2-3 s is really slow, that shouldn't be. The database stores its files on an SSD too? 67k photos isn't much. In any case, please increase the innodb cache as described in our docs.

**Reply by joachimtingvold** [reply]
@inthreedee wrote:
> You imply the work is already done, but I figured I'd plop this here in case it's useful to you.

Thanks, apprecaite it. I'm not done yet, but getting there. I knew there probably would be multiple tools already made to migrate to PP, but I wanted to "get my hands dirty" in terms of getting to know the inner workings of PP a bit more (since I've already made the decision to commit). I also had to "clean up" some old stuff, so it wouldn't be a 1:1 migration.

@lastzero wrote:
> Why not use SQL directly?

Because generally you want to use the API when there is one?

> Besides 2-3 s is really slow, that shouldn't be.

Doing `/api/v1/photos?count=1000&offset=0` (without `merged=true`), this seems to be the actual SQL query from PP;

```
SELECT photos.*, photos.id AS composite_id,
		files.id AS file_id, files.file_uid, files.instance_id, files.file_primary, files.file_sidecar,
		files.file_portrait,files.file_video, files.file_missing, files.file_name, files.file_root, files.file_hash,
		files.file_codec, files.file_type, files.file_mime, files.file_width, files.file_height,
		files.file_aspect_ratio, files.file_orientation, files.file_main_color, files.file_colors, files.file_luminance,
		files.file_chroma, files.file_projection, files.file_diff, files.file_duration, files.file_size,
		cameras.camera_make, cameras.camera_model,
		lenses.lens_make, lenses.lens_model,
		places.place_label, places.place_city, places.place_state, places.place_country
FROM `photos`
JOIN files ON photos.id = files.photo_id AND files.file_missing = 0 AND files.deleted_at IS NULL
LEFT JOIN cameras ON photos.camera_id = cameras.id
LEFT JOIN lenses ON photos.lens_id = lenses.id
LEFT JOIN places ON photos.place_id = places.id
WHERE (files.file_type = 'jpg' OR files.file_video = 1) AND (files.file_error = '') AND (photos.deleted_at IS NULL)
ORDER BY taken_at DESC, photos.photo_uid, files.file_primary DESC
LIMIT 1000 OFFSET 0

[…]

1000 rows in set (2.471 sec)
```

Running that manually in MariaDB takes about ~2.5s. If I throw in`merged=true`, it seems to do a shitload of individual queries from the `markers` table (probably for each stacked photo), but the initial query above seems to be the same.

I have about 134k entries in `files`, 76k in `photos` and 20k in `markers`.

> The database stores its files on an SSD too?

Yes.

> […] please increase the innodb cache as described in our docs.

I increased it to 1G from the get-go (i.e. before even trying out PP for the first time). I have since increased it to 4G, without any noticeable performance gain.

**Reply by joachimtingvold** [reply]
I copied the database to a different MariaDB-server I have (with more "oompf"), and re-ran the same query.

```
1000 rows in set (1.040 sec)
1000 rows in set (0.977 sec)
1000 rows in set (0.002 sec)
```

This MariaDB has query cache enabled, but even the initial query was almost 2.5 times faster. Maybe I have an excuse to buy better SSDs now :D

**Reply by lastzero** [reply]
Right... the query isn't fast and could be optimized in many ways, but also not that slow - even on our old, 2-core test server with "only" a consumer SATA SSD. If it's much slower than an old Core i3, you should look at your environment. Unfortunately, it's a lot of guesswork for us as to what the actual cause is.

Increased the file limit to 25000 for you with one of the recent commits. Development Preview has been updated too. Going to release it when tests are successful.

**For what do you need a list of all photos? Or is it files?** If it is something specific, you may only need very few columns and also only data from one of the tables without any joins.

I recommend to write a few lines of standard SQL instead of using the API, unless you want to make the code publicly available. This will be way faster and also we don't need to open the public API to DoS attacks. The API is also extremely detailed and contains many fields that you may never need. One of the reasons why the performance sucks for use cases where you need all results very quickly.

The upcoming Timeline API will be much faster, just like the existing GeoJSON API which however only returns photos with known or estimated location.

**Reply by joachimtingvold** [reply]
> Right... the query isn't fast and could be optimized in many ways, but also not that slow - even on our old, 2-core test server with "only" a consumer SATA SSD. If it's much slower than an old Core i3, you should look at your environment. Unfortunately, it's a lot of guesswork for us as to what the actual cause is.

No worries. Look at my comments more as observations. I don't expect you to debug my setup (=

> Increased the file limit to 25000 for you with one of the recent commits. Development Preview has been updated too. Going to release it when tests are successful.

Hihi, thanks.

> **For what do you need a list of all photos? Or is it files?** If it is something specific, you may only need very few columns and also only data from one of the tables without any joins.

The usecase right now was dealing with re-creating albums from my old solution. And since it was not a 1:1 conversion (i.e. I was merging some albums and whatnot), I basically wanted "all photos" from PP and old solution (stored in a Postgres database), so that I could search through them. I might have been able to do in a different way, but that was the first that came to mind (=

> I recommend to write a few lines of standard SQL instead of using the API

I would have if I had known beforehand that the API wasn't created for such things (=

> This will be way faster and also we don't need to open the public API to DoS attacks. The API is also extremely detailed and contains many fields that you may never need. One of the reasons why the performance sucks for use cases where you need all results very quickly.

Maybe GraphQL would be something for v2 of the API, then (=

**Reply by lastzero** [reply]
Done is better than perfect. GraphQL also has its drawbacks. You would first need to find a mature implementation for Go and find out what impact GraphQL has on the ORM/database schema as well as the frontend code. Could as well require developing or including a lot of additional code, which then has additional bugs and slows down development.
